tools/hidl/hidlcomp.cpp

```diff
@@ -7034,38 +7114,87 @@ void HIDLcompiler::Process()
         write_esp_ng();
         write_esp_ng_cpp();
     }
-    
+
     if (isSCM)
         write_clarion_HRPC_interfaces();
 }
 
 
+void HIDLcompiler::processExecutionProfiling()
+{
+    //
+    // metrics execution profiling requires the memory header. Search the services and/or methods to determine
+    // if enabled in any. If so, write the include statement
+    EspServInfo *si;
+    bool profilingEnabled = false;
+    StrBuffer tmp;
+    for (si=servs; si; si=si->next)
+    {
+        StrBuffer serviceProfilingOptions;
+        bool serviceProfileExecutionEnabled = si->getMetaStringValue(serviceProfilingOptions,"profile_execution");
+
+        //
+        // At the HIDL top level, execution profiling is enable if any service is enabled
+        executionProfilingEnabled |= serviceProfileExecutionEnabled;
+
+        //
+        // Go through each method and save any profile information to make if faster later when
+        // generating code. Note that the default for the method is either what is specified in the
+        // method or what is set at the service level. Still track the executionProfileEnabled flag for HIDL at the top
+        EspMethodInfo *mthi=NULL;
+        for (mthi=si->methods;mthi!=NULL;mthi=mthi->next)
+        {
+            //
+            // Collect method profile execution values
+            StrBuffer methodProfilingOptions;
+            bool methodProfileExecutionEnabled = mthi->getMetaStringValue(methodProfilingOptions, "profile_execution");
+            executionProfilingEnabled |= methodProfileExecutionEnabled;   // again, if a method is enabled, set top flag
+            if (serviceProfileExecutionEnabled || methodProfileExecutionEnabled)
+            {
+                if (!mthi->getMetaInt("no_profile_execution"))
+                {
+                    mthi->setEnableExecutionProfiling(true);
+                    if (!mthi->setExecutionProfilingOptions(methodProfileExecutionEnabled ? methodProfilingOptions.str() : serviceProfilingOptions.str()))
+                    {
+                        throw "Invalid execution profile options detected";
+                    }
+                }
+            }
+        }
+    }
+}
+
+
 void HIDLcompiler::write_esp()
 {
     //create the *.esp file
     gOutfile = espx;
     outf("// *** Source file generated by " HIDL " Version %s from %s.%s ***\n", HIDLVER, packagename, srcFileExt);
-    outf("// *** Not to be hand edited (changes will be lost on re-generation) ***\n\n"); 
-    
+    outf("// *** Not to be hand edited (changes will be lost on re-generation) ***\n\n");
+
     outf("#ifndef %s_ESPGEN_INCLUDED\n", packagename);
     outf("#define %s_ESPGEN_INCLUDED\n\n", packagename);
     outf("#include \"%s_esp.ipp\"\n", packagename);
+
+    if (executionProfilingEnabled)
+        outs("#include \"espcommon.hpp\"\n");
+        outs("#include \"jmetrics.hpp\"\n");
```

Comment:
Did you mean?:
```
   if (executionProfilingEnabled)
   {
        outs("#include \"espcommon.hpp\"\n");
        outs("#include \"jmetrics.hpp\"\n");
   }
```

Reply:
Yes, probably. Good catch, although not harmful. Better to fix for sure.